components/script/dom/gpu.rs

```diff
@@ -0,0 +1,168 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::compartments::InCompartment;
+use crate::dom::bindings::codegen::Bindings::GPUBinding::GPURequestAdapterOptions;
+use crate::dom::bindings::codegen::Bindings::GPUBinding::{self, GPUMethods, GPUPowerPreference};
+use crate::dom::bindings::error::Error;
+use crate::dom::bindings::refcounted::{Trusted, TrustedPromise};
+use crate::dom::bindings::reflector::{reflect_dom_object, DomObject, Reflector};
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::gpuadapter::GPUAdapter;
+use crate::dom::promise::Promise;
+use crate::task::TaskOnce;
+use dom_struct::dom_struct;
+use ipc_channel::ipc::{self, IpcSender};
+use ipc_channel::router::ROUTER;
+use js::jsapi::Heap;
+use std::rc::Rc;
+use webgpu::wgpu;
+use webgpu::{WebGPURequest, WebGPUResponse, WebGPUResponseResult, WebGPUThread};
+
+#[dom_struct]
+pub struct GPU {
+    reflector_: Reflector,
+}
+
+impl GPU {
+    pub fn new_inherited() -> GPU {
+        GPU {
+            reflector_: Reflector::new(),
+        }
+    }
+
+    pub fn new(global: &GlobalScope) -> DomRoot<GPU> {
+        reflect_dom_object(Box::new(GPU::new_inherited()), global, GPUBinding::Wrap)
+    }
+
+    fn get_wgpu_thread(&self) -> Option<WebGPUThread> {
+        self.global().as_window().webgpu_thread()
+    }
+}
+
+pub trait AsyncWGPUListener {
+    fn handle_response(&self, response: WebGPUResponse, promise: &Rc<Promise>);
+}
+
+struct WGPUResponse<T: AsyncWGPUListener + DomObject> {
+    trusted: TrustedPromise,
+    receiver: Trusted<T>,
+}
+
+impl<T: AsyncWGPUListener + DomObject> WGPUResponse<T> {
+    #[allow(unrooted_must_root)]
+    fn response(self, response: WebGPUResponseResult) {
+        // JSAutoRealm needs to be manually made.
+        // Otherwise, Servo will crash.
+        let promise = self.trusted.root();
+        match response {
+            Ok(response) => self.receiver.root().handle_response(response, &promise),
+            Err(error) => promise.reject_error(Error::Type(format!(
+                "Received error from WebGPU thread: {}",
+                error
+            ))),
+        }
+    }
+}
+
+pub fn response_async<T: AsyncWGPUListener + DomObject + 'static>(
+    promise: &Rc<Promise>,
+    receiver: &T,
+) -> IpcSender<WebGPUResponseResult> {
+    let (action_sender, action_receiver) = ipc::channel().unwrap();
+    let task_source = receiver.global().networking_task_source();
+    let mut trusted = Some(TrustedPromise::new(promise.clone()));
+    let trusted_receiver = Trusted::new(receiver);
+    ROUTER.add_route(
+        action_receiver.to_opaque(),
+        Box::new(move |message| {
+            struct ListenerTask<T: AsyncWGPUListener + DomObject> {
+                context: WGPUResponse<T>,
+                action: WebGPUResponseResult,
+            }
+
+            impl<T> TaskOnce for ListenerTask<T>
```

Comment:
By the way you can use the `task!` macro instead, see https://github.com/servo/servo/blob/2caa227e10369c4590cba5b6bceb85d907cd1f8c/components/script/dom/globalscope.rs#L234

Reply:
Added it here: https://github.com/servo/servo/pull/24708/commits/5f06132f65fb9ff7dd6a84e4a2eaa0c118ad76ea#diff-7a9976d0e2083724e30717170c8df321R96